ast/compile.go

```diff
@@ -1728,6 +1735,218 @@ func (c *Compiler) rewriteDynamicTerms() {
+	// Only parse annotations if rego.metadata built-ins are called
+	regoMetadataCalled := false
+	for _, name := range c.sorted {
+		mod := c.Modules[name]
+		WalkRules(mod, func(rule *Rule) bool {
+			WalkExprs(rule, func(expr *Expr) bool {
+				if isRegoMetadataChainCall(expr) || isRegoMetadataRuleCall(expr) {
+					regoMetadataCalled = true
+				}
+				return regoMetadataCalled
+			})
+			return regoMetadataCalled
+		})
+
+		if regoMetadataCalled {
+			break
+		}
+	}
+
+	if regoMetadataCalled {
+		// NOTE: Possible optimization: only parse annotations for modules on the path of rego.metadata-calling module
+		for _, name := range c.sorted {
+			mod := c.Modules[name]
+
+			if len(mod.Annotations) == 0 {
+				var errs Errors
+				mod.Annotations, errs = parseAnnotations(mod.Comments)
+				errs = append(errs, attachAnnotationsNodes(mod)...)
+				for _, err := range errs {
+					c.err(err)
+				}
+			}
+		}
+	}
+}
+
+func (c *Compiler) rewriteRegoMetadataCalls() {
+	for _, name := range c.sorted {
+		mod := c.Modules[name]
+
+		WalkRules(mod, func(rule *Rule) bool {
+			var regoMetadataChainCalled bool
+			var regoMetadataRuleCalled bool
+
+			WalkExprs(rule, func(expr *Expr) bool {
+				if isRegoMetadataChainCall(expr) {
+					regoMetadataChainCalled = true
+				} else if isRegoMetadataRuleCall(expr) {
+					regoMetadataRuleCalled = true
+				}
+				return regoMetadataChainCalled && regoMetadataRuleCalled
+			})
+
+			if regoMetadataChainCalled || regoMetadataRuleCalled {
+				body := make(Body, 0, len(rule.Body)+2)
+
+				var metadataChainVar Var
+				if regoMetadataChainCalled {
+					// Create and inject metadata chain for rule
+
+					chain, err := createMetadataChain(c.annotationSet.Chain(rule))
+					if err != nil {
+						c.err(err)
+						return false
+					}
+
+					f := newEqualityFactory(c.localvargen)
+					eq := f.Generate(chain)
+					metadataChainVar = eq.Operands()[0].Value.(Var)
+					body.Append(eq)
+				}
+
+				var metadataRuleVar Var
+				if regoMetadataRuleCalled {
+					// Create and inject metadata for rule
+
+					var metadataRuleTerm *Term
+
+					a := getPrimaryRuleAnnotations(c.annotationSet, rule)
+					if a != nil {
+						annotObj, err := a.toObject()
+						if err != nil {
+							c.err(err)
+							return false
+						}
+						metadataRuleTerm = NewTerm(*annotObj)
+					} else {
+						// If rule has no annotations, assign an empty object
+						metadataRuleTerm = ObjectTerm()
+					}
+
+					f := newEqualityFactory(c.localvargen)
+					eq := f.Generate(metadataRuleTerm)
+					metadataRuleVar = eq.Operands()[0].Value.(Var)
+					body.Append(eq)
+				}
+
+				for _, expr := range rule.Body {
+					body.Append(expr)
+				}
+				rule.Body = body
+
+				vis := func(b Body) bool {
+					for _, err := range rewriteRegoMetadataCalls(metadataChainVar, metadataRuleVar, b, &c.RewrittenVars) {
+						c.err(err)
+					}
+					return false
+				}
+				WalkBodies(rule.Head, vis)
+				WalkBodies(rule.Body, vis)
```

Comment:
Hmm some questions:

1. Is `WalkBodies(rule.Head, vis)` not a no-op? Can there be ast.Body in a rule.Head?
2. Related, do these work right now?
```rego
r = rego.metadata.rule() { true }
s[rego.metadata.rule()] { true }
t[k] = rego.metadata.rule() { k := "rule" }
```

Reply:
1. `p = [x | x := rego.metadata.chain()]` would be an example of where there is a body in the rule head.
2. Yes:
```rego
# METADATA
# title: R
r = rego.metadata.rule() { true }

# METADATA
# title: S
s[rego.metadata.rule()] { true }

# METADATA
# title: T
t[k] = rego.metadata.rule() { k := "rule" }
```
yields
```json
{
"r": {
    "scope": "rule",
    "title": "R"
  },
  "s": [
    {
      "scope": "rule",
      "title": "S"
    }
  ],
  "t": {
    "rule": {
      "scope": "rule",
      "title": "T"
    }
  }
}
```

Reply:
Thanks!